x-pack/plugin/indices-write-load-tracker/src/main/java/org/elasticsearch/xpack/indiceswriteloadtracker/IndicesWriteLoadStatsCollector.java

```diff
@@ -0,0 +1,250 @@
+                    // on a cluster state applier thread.
+                    () -> getParentDataStreamName(shardId.getIndexName()),
+                    indexShard,
+                    indexShard.routingEntry().primary(),
+                    relativeTimeInNanosSupplier
+                )
+            );
+        }
+    }
+
+    @Override
+    public void afterIndexShardClosed(ShardId shardId, IndexShard indexShard, Settings indexSettings) {
+        histograms.remove(shardId);
+    }
+
+    private void cleanRolledOverIndices() {
+        final var indicesLookup = clusterService.state().metadata().getIndicesLookup();
+        histograms.entrySet().removeIf(entry -> {
+            final var shardId = entry.getKey();
+            return isDataStreamWriteIndex(shardId.getIndex(), indicesLookup) == false;
+        });
+    }
+
+    private static class ShardWriteLoadHistogram {
+        private final Supplier<String> parentDataStreamNameSupplier;
+
+        private volatile String parentDataStreamName;
+
+        private final IndexShard indexShard;
+        private final boolean primary;
+        private final LongSupplier relativeTimeInNanosSupplier;
+        private final Object histogramsLock = new Object();
+        private final Histogram indexingTimeHistogram;
+        private final Histogram mergeTimeHistogram;
+        private final Histogram refreshTimeHistogram;
+        private long lastTotalIndexingTimeSample;
+        private long lastTotalMergeTimeSample;
+        private long lastTotalRefreshTimeSample;
+        private long lastSampleRelativeTimeInNanos;
+
+        ShardWriteLoadHistogram(
+            Supplier<String> parentDataStreamNameSupplier,
+            IndexShard indexShard,
+            boolean primary,
+            LongSupplier relativeTimeInNanosSupplier
+        ) {
+            this.parentDataStreamNameSupplier = parentDataStreamNameSupplier;
+            this.indexShard = indexShard;
+            this.primary = primary;
+            this.relativeTimeInNanosSupplier = relativeTimeInNanosSupplier;
+            this.indexingTimeHistogram = createHistogram();
+            this.mergeTimeHistogram = createHistogram();
+            this.refreshTimeHistogram = createHistogram();
+
+            this.lastSampleRelativeTimeInNanos = relativeTimeInNanosSupplier.getAsLong();
+            this.lastTotalIndexingTimeSample = indexShard.getTotalIndexingTimeInNanos();
+            this.lastTotalMergeTimeSample = indexShard.getTotalMergeTimeInNanos();
+            this.lastTotalRefreshTimeSample = indexShard.getTotalRefreshTimeInNanos();
+        }
+
+        void recordSample() {
+            long totalIndexingTimeInNanosSample = indexShard.getTotalIndexingTimeInNanos();
+            long totalMergeTimeInNanosSample = indexShard.getTotalMergeTimeInNanos();
+            long totalRefreshTimeInNanos = indexShard.getTotalRefreshTimeInNanos();
+
+            long sampleRelativeTimeInNanos = relativeTimeInNanosSupplier.getAsLong();
```

Comment:
I would find it nicer to only get the time once per sample and then use that throughout. Would that cause issues?